Python/_warnings.c

```diff
@@ -1000,8 +1000,10 @@ do_warn(PyObject *message, PyObject *category, Py_ssize_t stack_level,
                        &filename, &lineno, &module, &registry))
         return NULL;
 
+    Py_BEGIN_CRITICAL_SECTION(tstate->interp);
```

Comment:
We can't lock `tstate->interp` because it's not a `PyObject*`. I think we should lock something else? (I'm not sure what.)

Reply:
Aah, I misremembered that API as taking any pointer. Ok, let's find something else.

Reply:
Well, what about the per-interpreter dict?

Reply:
Locking the interpreter dict seems a bit awkward. I would lean towards:

1) Add a `PyMutex` to `WarningsState`
2) Add a macro to `pycore_critical_section.h` that takes a `PyMutex*` instead of a `PyObject*` (and is a no-op) in the default build.
3) Use the newly added macro

Reply:
Alternatively, put a mutex in `_warnings_runtime_state`.

Reply:
I think we'll want to use the critical section API rather than calling `PyMutex_Lock()` directly because `warn_explicit` can *probably* end up executing arbitrary code (due to things like destructors) and we want to avoid the risk of deadlocks.

Reply:
Another alternative is to create impl functions for the three helpers (`get_once_registry`, etc.), and protect those using critical sections.

Reply:
> Another alternative is to create impl functions for the three helpers (`get_once_registry`, etc.), and protect those using critical sections.

Proposed in commit a87d84c5ec

Reply:
> Locking the interpreter dict seems a bit awkward. I would lean towards:
> 
> 1. Add a `PyMutex` to `WarningsState`
> 2. Add a macro to `pycore_critical_section.h` that takes a `PyMutex*` instead of a `PyObject*` (and is a no-op in the default build).
> 3. Use the newly added macro

Would it not be overkill to add a dedicated critical section API just for this case?

Reply:
CI is failing, https://github.com/python/cpython/commit/a87d84c5ec60a41b4821bb4d80c4b366e1916b63 is apparently not enough.

Reply:
I'll give your idea from https://github.com/python/cpython/pull/116959#discussion_r1531220091 a try in a day or two.

Reply:
> I'll give your idea from [#116959 (comment)](https://github.com/python/cpython/pull/116959#discussion_r1531220091) a try in a day or two.

Well, last week was busy. I finally got around to give this a new try. See f9e549dd59c741b7f336560aa430e0d6c773b2c8. Bikeshedding for the internal API welcome.